### PR TITLE
rustdoc: Another round of improvements

### PR DESCRIPTION
**Pull request by alexcrichton**
A few final improvements to rustdoc for 1.0:
- Improve how rustdoc handles stability
- Fix cross-crate macro source links
- Improve experience of types inheriting methods through `Deref`

Some more details can be found in the commits.

[Preview](http://people.mozilla.org/~acrichton/doc/std/)

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @aturon 

cc @sfackler, @wycats, @steveklabnik, @bstrie, @tomjakubowski

**Comment by sfackler** [comment]
Yay! It'd be nice to identify deprecated stuff in search results as well.

**Comment by wycats** [comment]
r+ x 1000. It would be useful to get before and after screenshots for rustdoc changes :smile:

**Comment by tomjakubowski** [comment]
Wow, fantastic work as always Alex! I love the changes to stability and of course the "Methods from `Deref`" sections/

**Comment by killercup** [comment]
Wow, those unstable/deprecated boxes are huge! I'd probably make them look a bit less… alerting, e.g. by using a border-size of only 1px and not make the whole text italic. Might be good to add some margin-bottom to them, as well.

Oh, an here are some screenshots, for posterity.

std__io__charserror_-_rust
std__rand_-_rust

**Comment by steveklabnik** [comment]
I agree that overall this is an improvement, but agree with @killercup that it's a bit big.

**Comment by alexcrichton** [comment]
Yeah I definitely wouldn't slate myself as a graphic designer any time soon!

@wycats you can compare/contrast the current docs to the hosted docs (what I've been doing) for example:

| type  | old | new |
-------|--------|------
Vec | [old](http://doc.rust-lang.org/std/vec/struct.Vec.html) | [new](http://people.mozilla.org/~acrichton/doc/std/vec/struct.Vec.html)
old_io | [old](http://doc.rust-lang.org/std/old_io/index.html) | [new](http://people.mozilla.org/~acrichton/doc/std/old_io/index.html)
io | [old](http://doc.rust-lang.org/std/io/index.html) | [new](http://people.mozilla.org/~acrichton/doc/std/io/index.html)

**Comment by liigo** [comment]
Another great feature's screenshot, thanks to @alexcrichton !
`Vec` methods from `Deref`:

vec-deref-methods

**Comment by aturon** [comment]
This is really awesome work, @alexcrichton! The docs are feeling much more polished, and I really like the `Deref` treatment.

I do agree with others that the unstable/deprecation warnings are a bit too attention-getting. I suspect that some fairly simple changes to the styling could de-emphasize them, and that seems worth doing in this PR.

**Comment by alexcrichton** [comment]
Ok I've pushed a few tweaks to the style (and re-uploaded the docs). I incorporated @killercup's suggestions (thanks!) and also made the font size a little smaller. I do think that it's still a little jarring but much less so than before!

**Comment by steveklabnik** [comment]
:+1:

**Comment by killercup** [comment]
:+1: 

(The padding on the bottom looks a bit off to me, but that's nothing major. I'd probably set it to `5px 5px 2px` or something like that.)

**Comment by aturon** [comment]
Thanks @alexcrichton! The code looks good to me as well, so:

@bors: r+

**Comment by bluss** [comment]
Users will be overjoyed to see the Deref changes (well they should be), but DerefMut methods are still missing.

File: src/librustdoc/clean/mod.rs [hunk 1]
```diff
@@ -313,6 +317,22 @@ impl Item {
...
     }
+
+    pub fn stability_class(&self) -> String {
+        match self.stability {
+            Some(ref s) => {
+                let mut base = match s.level {
+                    attr::Unstable => "unstable".to_string(),
+                    attr::Stable => String::new(),
+                };
+                if s.deprecated_since.len() > 0 {
```
Comment (tamird) [inline src/librustdoc/clean/mod.rs]: @alexcrichton `if !s.deprecated_since.is_empty() {`?

File: src/librustdoc/html/render.rs [hunk 2]
```diff
@@ -1685,6 +1688,30 @@ fn item_module(w: &mut fmt::Formatter, cx: &Context,
     write!(w, "</table>")
 }
 
+fn short_stability(item: &clean::Item, show_reason: bool) -> Option<String> {
+    item.stability.as_ref().and_then(|stab| {
+        let reason = if show_reason && stab.reason.len() > 0 {
```
Comment (tamird) [inline src/librustdoc/html/render.rs]: and here

File: src/librustdoc/html/render.rs [hunk 3]
```diff
@@ -1685,6 +1688,30 @@ fn item_module(w: &mut fmt::Formatter, cx: &Context,
...
+            format!(": {}", stab.reason)
+        } else {
+            String::new()
+        };
+        let text = if stab.deprecated_since.len() > 0 {
```
Comment (tamird) [inline src/librustdoc/html/render.rs]: and here

File: src/librustdoc/clean/mod.rs (hunk 1 above)
Comment (tamird) [inline src/librustdoc/clean/mod.rs]: disregard! i hadn't noticed that the conflict is not yet resolved here

**Comment by Stebalien** [comment]
Nice. However, IMHO, the Deref docs shouldn't include static methods.

**Comment by tomjakubowski** [comment]
Probably worth filing an issue for that

**Comment by Stebalien** [comment]
@tomjakubowski 
> Probably worth filing an issue for that

Good point. Done.

**Comment by SimonSapin** [comment]
This is great! Thanks Alex.

> Improve experience of types inheriting methods through Deref 

The commit message says this is only inherent methods through Deref. Would it make sense to also show methods available through trait implementations on the Deref target?

**Comment by alexcrichton** [comment]
@SimonSapin my thinking was that the target of the `Deref` impl generally doesn't implement traits that the source itself does not. For example the traits implemented by `String` and `str` are almost the same, so I didn't think it was worth it to add a section for "trait implementations available as methods through `Deref`".

I chose to only inline inherent methods because that's the real benefit of `Deref` in terms of docs, you get new inherent methods you didn't previously have.

**Comment by SimonSapin** [comment]
Fair enough, especially now that primitive types can have inherent methods and `std` doesn’t have so many traits.

**Pull request merged**